packages/utils/telemetry-utils/src/logger.ts

```diff
@@ -49,32 +49,54 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
-        if (error === null || typeof error !== "object") {
+        if (typeof error !== "object" || error === null) {
             event.error = error;
         } else {
-            // WARNING: Exceptions can contain PII!
-            // For example, XHR will throw object derived from Error that contains config information
-            // for failed request, including all the headers, and thus - user tokens!
-            // Extract only call stack, message, and couple network-related properties form error object
-
-            const errorAsObject = error as {
-                stack?: string;
-                message?: string;
-            };
-
+            const errorAsObject = error as Partial<Error>;
             event.stack = errorAsObject.stack;
             event.error = errorAsObject.message;
 
-            // Error message can contain PII information.
-            // If we know for sure it does, we have to not log it.
-            if (error.containsPII) {
-                event.error = "Error message was removed as it contained PII";
-            } else if (error.getTelemetryProperties) {
-                const telemetryProps: ITelemetryProperties = error.getTelemetryProperties();
-                for (const key of Object.keys(telemetryProps)) {
-                    if (event[key] === undefined) {
-                        event[key] = telemetryProps[key];
+            if (isILoggingError(error)) {
+                const taggableProps = error.getTelemetryProperties();
+                for (const key of Object.keys(taggableProps)) {
+                    if (event[key] !== undefined) {
+                        continue; //* test this
+                    }
+                    const taggableProp = taggableProps[key];
+                    const { value, tag } = (typeof taggableProp === "object")
+                        ? taggableProp
+                        : { value: taggableProp, tag: TelemetryDataTag.None };
+                    switch (tag) {
+                        case TelemetryDataTag.None:
+                        case TelemetryDataTag.PackageData:
+                            // For Microsoft applications, PackageData is safe for now
+                            // (we don't load 3P code in 1P apps)
+                            // But this determination really belongs in the host layer
+                            event[key] = value;
+                            break;
+                        case TelemetryDataTag.UserData:
+                            // Strip out anything tagged explicitly as PII.
+                            // Alternate strategy would be to hash these props
+                            event[key] = undefined;
+                            break;
+                        default:
+                            // This will help us keep this switch statement up to date
+                            (function(_: never) {})(tag);
+
+                            // If we encounter a tag we don't recognize
```

Comment:
That's not the only possible reason. Someone can just put a complex object as a property. Type does not allow that, but casting to any is easy solution :) Should we somehow warn developers?
Note - mismatch in versions is definitely a possibility (with errors from driver & runtime reported by loader). But it's trivial to detect - we can make LoggingError class report "telemetry version" it's from (i.e. version of package where LoggingError class is), and thus this code can easily figure out if it comes from future or past (only future is a problem).

Reply:
I tried to block the complex object problem in `getValidTelemetryProperties` using this type guard:

```
export function isTaggedTelemetryPropertyValue(x: any): x is ITaggedTelemetryPropertyType {
    return (typeof(x?.value) !== "object" && typeof(x?.tag) === "string");
}
```

It's not perfect but in practice pretty close.  When we encounter an object that doesn't meet that criteria, how about we still include the key but put a canned message about the object being redacted.

I'm opening an issue to track the version-tracking suggestion, we can discuss more there.

Reply:
https://github.com/microsoft/FluidFramework/issues/5544